### PR TITLE
refactor: move TransportSocketOptions into LoadBalancerContext

### PR DESCRIPTION
**Pull request by lizan**
Description:
Needed for #2690 to use TransportSocketOptions in HTTP cluster.

Risk Level: Low (refactor only)
Testing: existing
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: include/envoy/upstream/load_balancer.h [hunk 1]
```diff
@@ -76,6 +77,11 @@ class LoadBalancerContext {
    * Returns the set of socket options which should be applied on upstream connections
    */
   virtual Network::Socket::OptionsSharedPtr upstreamSocketOptions() const PURE;
+
+  /**
+   * Returns the transprot socket options which should be applied on upstream connections
```
Comment (snowp) [inline include/envoy/upstream/load_balancer.h]: typo: transport

File: source/common/upstream/cluster_manager_impl.cc [hunk 2]
```diff
@@ -1254,16 +1253,18 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::tcpConnPool(
     }
   }
 
-  if (transport_socket_options != nullptr) {
-    transport_socket_options->hashKey(hash_key);
+  bool have_transport_socket_options = false;
+  if (context && context->upstreamTransportSocketOptions() != nullptr) {
```
Comment (snowp) [inline source/common/upstream/cluster_manager_impl.cc]: nit: `context != nullptr`

File: include/envoy/upstream/load_balancer.h (hunk 1 above)
Comment (zuercher) [reply include/envoy/upstream/load_balancer.h]: Weird that the spell checker didn't find this.

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 3]
```diff
@@ -964,16 +964,16 @@ TEST_F(ClusterManagerImplTest, UnknownCluster) {
   EXPECT_EQ(nullptr, cluster_manager_->httpConnPoolForCluster("hello", ResourcePriority::Default,
                                                               Http::Protocol::Http2, nullptr));
   Network::TransportSocketOptionsSharedPtr transport_socket_options;
```
Comment (zuercher) [inline test/common/upstream/cluster_manager_impl_test.cc]: I think this is unused now and can be removed.

**Pull request merged**